### ISSUE TITLE
Redo

### ISSUE THREAD

**Issue by andreasdc** [eclipse-jdt/eclipse.jdt.ui#454]

**Comment by JanecekPetr** [comment]
This is pretty standard behaviour in any editor as it creates a fork in the history and Ctrl+Y then does not know which way to take in the history. What do you suggest would be the ideal alternative?

You can still right-click your file and select Team -> Show Local History to see all (?) your previous versions.

**Comment by andreasdc** [comment]
this is not normal

**Comment by JanecekPetr** [comment]
> this is not normal

I'm sorry, you will have to be more specific than this as I apparently do not understand the problem. I just tried this in my OS's text editor:

1. I wrote a few sentences.
2. I pressed the Undo, some of the words disappeared.
3. I pressed Redo, everything came back.
4. I pressed Undo again, making some words disappear.
5. I made a "mistake" by pressing the letter 'z' instead of pressing the shortcut for Redo.
6. From this point on, no combination of Undo or Redo operations showed me the original sentences, I only could see the letter 'z' appear and disappear.

This seems consistent with what you described as being the problem. If it is, sorry, that's how standard text editors work. If I misunderstood the issue, please describe it in more detail so that someone from the Eclipse JDT team can understand.

Note that the Local History workaround should still work.

**Comment by JanecekPetr** [comment]
Some more context here: https://github.com/eclipse-platform/.github/issues/99

I just tried Notepad++ and it behaves as expected, just like eclipse / Word / any other standrd text editor does. If you think otherwise, please show us a way to reproduce it. Other than that, the Local History feature will be your friend, and obviously a versioning system like Git.

**Comment by andreasdc** [comment]
Notepad++ doesn't delete my code, while in Eclipse 1 missclicked button and you lose your code.

**Comment by JanecekPetr** [comment]
Please, help us understand what you mean by showing us what you're using in notepad++. I don't know it. I tried it, did the same sequence of operations as mentioned above in my OS's text editor (TextEdit on Mac) experiment, and got the same result.

Is there any other kind of operation you're using besides Undo and Redo? If yes, which one and how do I reproduce what you're seeing? Does notepad++ have some kind of extra local history that you can get back to? If yes, how is it different/better from Eclipse's Local History feature?

Of course, in any serious project Git (or other versioning system) is the answer which is why it's unlikely a change will happen to eclipse's history and/or Local History features. Still, I'm curious. Maybe you did actually find a bug and we just don't yet understand what you mean.

**Comment by andreasdc** [comment]
But you know on eclipse after you do redo, any kind of key pressed will delete your undo history? No matter if it's typed letter or button that won't play any action.

**Comment by JanecekPetr** [comment]
> But you know on eclipse after you do redo, any kind of key pressed will delete your undo history? No matter if it's typed letter or button that won't play any action.

No, not any key press. If you are in the middle of the undo/redo stack and press a key which does an _Edit_, then and only then it will add your last action on the stack and forget the original top of the stack. I just tried, even actions that _could_ do an edit but don't (like Optimize Imports on an already good file or a cancelled Content Assist) are okay, Eclipse still remembers the whole history chain.

As we said countless times, everything we observe in Eclipse regarding Undo/Redo is exactly the same as in every text editor we've seen, including notepad++. Please, if you're seeing something else, give us a step-by-step reproducible example. Or record a video. We clearly do not understand each other.

**Comment by andreasdc** [comment]
For me pressing any key other than redo removes the h istory.

**Comment by JanecekPetr** [comment]
Aha! Now that sounds like a bug, I think. Would you mind sending your eclipse log file? It should be located at .../your-workspace-location/.metadata/.log. Is there anything interesting / relevant there? Try triggering the issue, then look into the file.

**Comment by carstenartur** [comment]
To investigate the reported behaviour in a reproducible way, I created an investigation-only SWTBot test suite that makes the document-specific Undo/Redo history directly observable instead of checking only the visible editor text.

The test code and the successful upstream CI run are available in PR #3147⁠￼. The corrected run completed in the normal JDT UI Jenkins build with 9,436 tests passed, 154 skipped, and no failures.

The investigation covers:

* different positions within the Undo/Redo history, including a remaining Redo branch;
* the Java editor and the generic text editor;
* native AWT/Robot input and SWT-injected Ctrl+H;
* navigation keys and modifier-only key presses;
* Backspace and Delete when they do not change the document;
* Save and the platform-specific Redo shortcut;
* positive controls that perform real text edits.

In addition to the visible text, the tests compare the order and identity of the Undo and Redo operations, the document modification stamp, document events, operation-history events, and the enabled state of the Undo and Redo actions. After each non-editing interaction, Redo is actually executed to verify that the previous document state can still be restored.

For the tested current master configuration under Linux/GTK, the result is:

* A real edit, such as typing a character or deleting text, creates a new history branch and therefore discards the previous Redo branch. This is the expected behaviour of a linear Undo model.
* Ctrl+H, navigation keys, modifier-only key presses, Save, and deletion keys that do not actually modify the document preserve the complete Undo/Redo history. Redo remains available and restores the expected text afterwards.

Based on this result, the broader statement that any key press removes the history is not reproducible on current master in this environment.

This does not necessarily disprove an environment-specific problem in an older Eclipse version, on Windows or macOS, with a particular keyboard layout or key scheme, with installed plug-ins, or in a specific workspace configuration. Reproducing such a case would require the corresponding environment details and a more precise sequence of actions.

PR #3147 is deliberately not intended for merging in its current form; it was created as a diagnostic experiment. If anyone has another plausible scenario, a different interpretation of the original report, or a better testing approach, please feel free to reuse or extend the code from the PR. The included diagnostics should help distinguish between a real document edit, removal of a history operation, an incorrect action state, and an interaction that leaves the stack unchanged.

In the absence of a more specific reproducer, closing this issue as currently not reproducible would seem reasonable, while the diagnostic code remains available as a basis for further investigation.

**Comment by andreasdc** [comment]
Sounds cool! How to use it?

**Comment by carstenartur** [comment]
> Sounds cool! How to use it?

Just to clarify: this is not a user-facing Eclipse feature or an Undo Stack view that can be installed. It is an investigation-only test harness for Eclipse contributors.

The code is available in PR #3147:
https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/3147

The main test class is here:
https://github.com/carstenartur/eclipse.jdt.ui/blob/investigate-454-undo-stack-swtbot/org.eclipse.jdt.ui.swtbot.tests/src/org/eclipse/jdt/ui/swtbot/tests/UndoRedoHistorySWTBotTest.java

The CI-specific adapter is here:
https://github.com/carstenartur/eclipse.jdt.ui/blob/investigate-454-undo-stack-swtbot/org.eclipse.jdt.ui.swtbot.tests/src/org/eclipse/jdt/ui/swtbot/tests/UndoRedoHistorySWTBotCITest.java

A focused Linux test run can be started from the repository root with:

xvfb-run -a mvn -B -ntp \
  -Pbuild-individual-bundles \
  -Dcompare-version-with-baselines.skip=true \
  -pl org.eclipse.jdt.ui.swtbot.tests -am \
  verify

On Windows or macOS, omit the "xvfb-run -a" part. Native keyboard tests should preferably run in a real interactive desktop session.

To investigate another scenario, add a test method that:

1. prepares an Undo/Redo state,
2. performs the interaction to be investigated,
3. compares the document and history before and after the interaction,
4. performs Redo afterwards to verify that the previous state can still be restored.

On failure, the test reports information such as:

- the document contents and modification stamp,
- the Undo and Redo operations,
- operation identities and contexts,
- document events,
- operation-history events,
- keyboard events,
- Undo/Redo action states,
- the active editor, focus and open shells.

SWTBot is useful when the suspected problem depends on the complete Eclipse UI interaction, for example:

- key-binding resolution,
- editor focus,
- active windows or dialogs,
- StyledText keyboard handling,
- native versus SWT-injected input,
- interactions that cannot be reproduced through document APIs alone.

SWTBot is usually not the best first choice for testing the Undo model or operation-history logic in isolation. Tests directly against IDocumentUndoManager, IOperationHistory or a document implementation are faster, more deterministic and easier to run across platforms.

A sensible test strategy is therefore:

- focused lower-level tests for the Undo/Redo logic,
- a small number of SWTBot tests for the complete UI integration.

It is also important that SWTBot only verifies the environment in which it runs. A suspected Windows-, macOS-, keyboard-layout-, key-scheme-, plug-in- or workspace-specific issue should be tested in that corresponding environment.

Further information:

SWTBot:
https://github.com/eclipse-swtbot/org.eclipse.swtbot

Tycho testing documentation:
https://tycho.eclipseprojects.io/doc/latest/TestingBundles.html

JDT contribution guide:
https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md

From my side, this completes the investigation. The reported general behaviour was not reproducible on current master under Linux/GTK, while the positive controls confirmed that the tests do detect a Redo branch being discarded after a real edit.

I do not currently plan to extend the investigation further without a concrete new reproducer or a specific environment-dependent hypothesis. However, the PR and diagnostic code remain available, and anyone with another plausible scenario or a better testing approach is very welcome to reuse and extend them.
